hkmc2Benchmarks/src/test/bench/mlscript-compile/nofib-deforest/mate.mls

```diff
@@ -0,0 +1,557 @@
+    emptyBoard
+    concat(zipWith(parseRank, reverse(enumFromTo(1, 8)), ls))
+
+fun parseProblem(s) =
+  let bdtxt_gltxt = splitAt(8, filter(x => not(comment(x)), s))
+  if bdtxt_gltxt is [bdtxt, gltxt] then
+    let bd = parseBoard(bdtxt)
+    let gl = parseGoal(gltxt)
+    [bd, gl]
+
+fun readProblem(s) = parseProblem(lines(s))
+
+fun foldr_lz(f, a, x) = if x is
+  h :: t then f(h, lazy of () => foldr_lz(f, a, t))
+  Nil then a
+
+//│ ————————————————————————————————————————————————————————————————————————————————
+fun replies(bd, c, n) =
+  let mds = moveDetailsFor(c, bd)
+  fun solnAnd(mifb, rest) = if mifb is [mif, b] then
+    let sm = solution(b, opponent(c), n - 1)
+    if sm is
+      None then None
+      Some(s) and
+        force(rest) is
+          None then None
+          Some(ms) then Some([mif, s] :: ms)
+  if n
+    === 0 and
+      null_(mds) then Some(Nil)
+      else None
+    > 0 then foldr_lz(solnAnd, Some(Nil), mds)
+    else throw Error("n < 0")
+
+fun solution(bd, c, n) =
+  fun solnOr(mifb, other) = if mifb is
+    [mif, b] then
+      let rsm = replies(b, opponent(c), n - 1)
+      if rsm is
+        None then force(other)
+        Some(Nil) and
+          kingincheck(opponent(c), b) then
+            Some(Solution(mif, Nil))
+          else force(other)
+        Some(rs) then Some(Solution(mif, rs))
+  if n > 0 then
+    let mds = moveDetailsFor(c, bd)
+    foldr_lz(solnOr, None, mds)
+  else throw Error("n <= 0")
+//│ ————————————————————————————————————————————————————————————————————————————————
+
+data class Soln(a: MoveInFull, b: List[[List[MoveInFull],Soln]])
+
+fun tab(n) = if n <= 0 then Nil else " " :: tab(n - 1)
+
+//│ ————————————————————————————————————————————————————————————————————————————————
+fun showReplies(rs, n) = if rs is
+  Nil then Nil
+  [mifs, s] :: rs then
+    tab(n) +:
+    nofibStringToList("if ") +:
+    if null_(rs) && (listLen(mifs) > 1) then nofibStringToList("others") else showMoves(mifs) +:
+    nofibStringToList("; ") +:
+    showSoln(s, n + 1) +:
+    showReplies(rs, n)
+
+fun showSoln(s, n) = if s is Soln(mif, rs) then
+  nofibStringToList(stringOfInt(n)) +:
+  nofibStringToList(". ") +:
+  showMoveInFull(mif) +:
+  if rs is
+    Nil then nofibStringToList("++\n")
+    [mifs, s_] :: Nil then
+      nofibStringToList(", ") +:
+      (if listLen(mifs) > 1 then nofibStringToList("...") else showMoves(mifs)) +:
+      nofibStringToList("; ") +:
+      showSoln(s_, n + 1)
+    else
+      nofibStringToList(",\n") +: showReplies(sort(rs), n)
+//│ ————————————————————————————————————————————————————————————————————————————————
+
+
+//│ ————————————————————————————————————————————————————————————————————————————————
+fun compact(s) = if s is Solution(mif, rs) then Soln(mif, foldr(insertCompact, Nil, rs))
+
+fun insertCompact(mif_s, ls) = if mif_s is [mif, s] then
+  fun insert(x, ls) = if ls is
+    Nil then x :: Nil
+    y :: ys then if x > y then y :: insert(x, ys) else x :: y :: ys
```

Comment:
Here `x > y` converts both x and y (which are JS objects of class `MoveInFull`) to string and perform string comparison. It should compare field by field instead